### ISSUE TITLE
postgres.upstream fails when installing only postgres.client

### ISSUE THREAD

**Issue by nmistry** [saltstack-formulas/postgres-formula#58]
https://github.com/saltstack-formulas/postgres-formula/blob/master/postgres/upstream.sls#L11-L12

`postgres.upstream` references the state defined by the variable `postgres.pkg`.  This state does not exist when installing  `postgres.client`.  

```
Cannot extend ID 'postgresql-9.4' in 'base:postgres.upstream'. It is not part of the high state.
This is likely due to a missing include statement or an incorrectly typed ID.
Ensure that a state with an ID of 'postgresql-9.4' is available
in environment 'base' and to SLS 'postgres.upstream'
```

**Comment by madssj** [comment]
Hey @iggy would this not be solved by removing the `require_in` statement? Seeing as salt should take care of the proper ordering anyway.

If not, it could be solved with implicit ordering.

I'd be happy to send a PR, just let me know how it'd be best solved here.

**Comment by iggy** [comment]
require_in on pkgrepo is a special case. It makes sure the package manager does an update (apt-get update) before it tries to install the package (vs just running the state).

It should probably be `- pkg: install-postgresql`

**Comment by myii** [comment]
Using the following workaround to avoid this but still pull all relevant packages from the upstream repo.  Slightly more verbose, unfortunately.

##### Workaround

Remove or comment out lines in `upstream.sls`:

```
install-postgresql-repo:
  ...
    # - require_in:
    #   - pkg: install-postgresql
```

Use conditional `require` in `init.sls`:

```
install-postgresql:
  ...
{% if postgres.use_upstream_repo %}
    - refresh: {{ postgres.use_upstream_repo }}
    - require:
      - pkgrepo: install-postgresql-repo
{% endif %}
```

Same again in `client.sls`:

```
install-postgresql-client:
  ...
{% if postgres.use_upstream_repo %}
    - refresh: {{ postgres.use_upstream_repo }}
    - require:
      - pkgrepo: install-postgresql-repo
{% endif %}
```

And that does the trick.  Note that the `if` could come after the `refresh` but have enclosed since the `refresh` and `require` are related logically via. `use_upstream_repo`.

##### Extra + libpg-dev

Probably unnecessary but to explicitly ensure that extra packages and libpq-dev are also pulled from the refreshed upstream repo (without triggering another refresh), have used `require` accordingly:

`init.sls`:

```
install-postgres-extra:
  ...
    - require:
      - pkg: install-postgresql
```

`client.sls`:

```
install-postgres-libpq-dev:
  ...
    - require:
      - pkg: install-postgresql-client
```

##### Ideal resolution

Would love to find out how this could be achieved more succinctly.  Not well-versed in SaltStack but something like the following could be the first step to a nicer resolution:

`upstream.sls`:

```
install-postgresql-repo:
  ...
    - require_in:
{% if postgres.client %}
      - pkg: install-postgresql-client
      - pkg: install-postgres-libpq-dev
{% else %}
      - pkg: install-postgresql
      - pkg: install-postgres-extra
{% endif %}
```

Is there any way to reference the state that is running?

**Comment by vutny** [comment]
@gravyboat This is finally resolved now after PR #124 was merged.
